Firestore/core/src/firebase/firestore/util/status_grpc.cc

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/status.h"
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+Status Status::FromGrpcStatus(const grpc::Status& from) {
+  if (from.ok()) {
+    return util::Status::OK();
+  }
+
+  grpc::StatusCode error_code = from.error_code();
+  HARD_ASSERT(
+      error_code >= grpc::CANCELLED && error_code <= grpc::UNAUTHENTICATED,
+      "Unknown gRPC error code: %s", error_code);
+
+  return {static_cast<FirestoreErrorCode>(error_code), from.error_message()};
```

Comment:
This relies on:
a) The integer values of the grpc statuscode enum
b) The integer values of our status code
c) These two integer values being the same.

That makes me nervous. Since their both based on Google's canonical error codes, this seems less bad though.

The alternative is a fairly massive switch statement. We do this for translations with errno, and with windows error codes (where we have no choice). That would be the safer approach here too, albeit much more verbose.

Switching to switch optional. (And I see it wasn't like that before, so this is certainly no worse.)

Reply:
This is pretty much taken from how `FSTDatastore` [does the conversion](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Remote/FSTDatastore.mm#L133-L136) currently. I think it's very unlikely that gRPC status would add non-canonical error codes, and on our side, we control the enum, so it should be okay.

Reply:
Update: added a comment.